pennylane/ops/qubit/observables.py

```diff
@@ -427,13 +427,25 @@ def __copy__(self):
 
         return copied_op
 
+    def __reduce__(self):
+        """Defines how to pickle and unpickle a :class:`~.Projector`. Circumbents the pickling
+        issues caused by its dynamic type:
+        >>> qml.Projector([1], wires=[0]) is qml.Projector
+        False
+        >>> isinstance(qml.Projector([1], wires=[0]), qml.Projector)
+        True
+        For more information, see: https://docs.python.org/3/library/pickle.html#object.__reduce__
+        """
+        state = self.data[0]
+        return (Projector, (state, self.wires), {"_id": self.id})
+
 
 class _BasisStateProjector(Observable):
     # The call signature should be the same as Projector.__new__ for the positional
     # arguments, but with free key word arguments.
     def __init__(self, state, wires, id=None):
         wires = Wires(wires)
-        state = list(qml.math.toarray(state))
+        state = list(qml.math.toarray(state).astype(int))
```

Comment:
I casted the input state to `int` because I found it annoying to not be able to directly pass an `np.zeros` or `np.ones` without setting them to `int` type. However, this may lead to wrong outputs silently when the input state is not supposed to be `int`. Previously it would break with a rather cryptic error message for the average user. 

I'd be happy with either: 
- sticking with this (potentially add a warning?)
- assess that the input state is of type `int` and provide an understandable error otherwise.

Reply:
when is the input state not supposed to be an int?

Reply:
In basis state projectors it's always supposed to be an int (0s or 1s). The issue is that, some times, we may have an array of int values but not int dtype like `[1.0, 0.0, 0.0]` and this will break the specialized branches for `expval` and `var` in lines such as [this one](https://github.com/PennyLaneAI/pennylane/blob/3fb3e5ee2063f59c6b20278c487cc1342f45f008/pennylane/_qubit_device.py#L1713), or in its `compute_matrix` method. I figured it'd be better to cast it as int at the beginning rather than everywhere else.

Reply:
I see. I would say it's still "supposed to" be an int in those cases, it just isn't because it was cast/defined as a float/complex. I'm good to cast at the beginning 👌

Reply:
once again, dtypes cause headaches.